### PR TITLE
Add ReplicatedParameter and ReplicatedTensor to DDP.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74787

As per the design in https://github.com/pytorch/pytorch/issues/72138,
introduce ReplicatedParameter and replace DDP parameters with
ReplicatedParameters and DDP buffers as ReplicatedTensors.

Differential Revision: [D35162813](https://our.internmc.facebook.com/intern/diff/D35162813/)

### CODE REVIEW

File: torch/distributed/_shard/replicated_tensor.py [hunk 1]
```diff
@@ -123,3 +123,27 @@ def validate(self) -> bool:
                     f"ReplicatedTensor have different values on rank {current_rank} and {rank}")
 
         return True
+
+class ReplicatedParameter(ReplicatedTensor, torch.nn.Parameter):
```
Comment (pritamdamania87) [inline torch/distributed/_shard/replicated_tensor.py:127]: @albanD I couldn't pass in a ReplicatedTensor to `torch.nn.Parameter` and have `isinstance()` still return ReplicatedTensor. As a result, I had to do something messy like this. Not sure if there is a better way to do this.

**Review by mrshenli — COMMENTED** [review]
Is this a test or do we plan to land this soon? If it's the latter, shall we do some micro-benchmarks to make sure there is no perf regression?

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -665,6 +665,16 @@ def _sync_params_and_buffers(self, authoritative_rank=0):
                 module_states, self.broadcast_bucket_size, authoritative_rank
             )
 
+        # After syncing, tag them as replicated.
+        from torch.distributed._shard.replicated_tensor import ReplicatedTensor, ReplicatedParameter
```
Comment (kumpera) [inline torch/nn/parallel/distributed.py:669]: Shouldn't we check if each `param` or `buffer` is replicated before converting them?
Similarly, shouldn't we record which ones we wrapped and use that when unwrapping, instead of relying on `parameters_to_ignore`.

Another thing, is using `parameters_to_ignore` to filter buffers just a naming artifact?

File: torch/distributed/_shard/replicated_tensor.py (hunk 1 above)
Comment (albanD) [reply torch/distributed/_shard/replicated_tensor.py:127]: cc @jbschlosser

File: torch/nn/parallel/distributed.py (hunk 2 above)
Comment (pritamdamania87) [reply torch/nn/parallel/distributed.py:669]: > Similarly, shouldn't we record which ones we wrapped and use that when unwrapping, instead of relying on parameters_to_ignore.

So `parameters_to_ignore` is slightly orthogonal to tagging something as replicated or not since that option is actually controlling which parameters DDP ignores completely. That is why we always need to rely on it and ensure we don't tag `parameters_to_ignore` as Replicated at all.

> Another thing, is using parameters_to_ignore to filter buffers just a naming artifact?

It's an unfortunate name since it is filtering buffers as well in other places. I'll probably fix this name in a follow up PR.

**Comment by pritamdamania87** [comment]
@mrshenli The plan is to hopefully land this change, however I think there are some tricky cases to take care of first. I'll probably ensure I can make CI happy first (tagged the PR as WIP) and then try a benchmark as well to ensure no regressions.

File: test/distributed/_shard/test_replicated_tensor.py [hunk 3]
```diff
@@ -74,3 +76,71 @@ def test_replicated_tensor_inter_op_tensor(self):
...
+            self.assertIsInstance(ddp.module.foo, ReplicatedTensor)
+            self.assertIsInstance(module.foo, ReplicatedTensor)
+
+        model.register_forward_pre_hook(hook)
+        for _ in range(2):
+            out = ddp(replica_tensor)
+            self.assertIsInstance(out, ReplicatedTensor)
+
+            # Validate buffers not replicated after forward.
+            self.assertNotIsInstance(ddp.module.foo, ReplicatedTensor)
```
Comment (rohan-varma) [inline test/distributed/_shard/test_replicated_tensor.py:145]: Don't we synchronize the buffers so it should be replicated?

File: test/distributed/_shard/test_replicated_tensor.py [hunk 4]
```diff
@@ -74,3 +76,71 @@ def test_replicated_tensor_inter_op_tensor(self):
...
+        # Test module.parameters.
+        params = list(ddp.parameters())
+        self.assertEqual(2, len(params))
+        self.assertEqual(ddp.module.weight, params[0])
+        self.assertEqual(ddp.module.bias, params[1])
+
+        params = list(model.parameters())
+        self.assertEqual(2, len(params))
+        self.assertEqual(model.weight, params[0])
+        self.assertEqual(model.bias, params[1])
```
Comment (rohan-varma) [inline test/distributed/_shard/test_replicated_tensor.py:103]: should we add `self.assertTrue(all(isinstance(replicatedParam, param) for param in model.parameters())`?

File: torch/nn/parallel/distributed.py [hunk 5]
```diff
@@ -601,11 +601,14 @@ def __init__(
         self.require_backward_grad_sync = True
         self.require_forward_param_sync = True
         self.gradient_as_bucket_view = gradient_as_bucket_view
-        if hasattr(module, "_ddp_params_and_buffers_to_ignore"):
-            self.parameters_to_ignore = module._ddp_params_and_buffers_to_ignore
+        if hasattr(self.module, "_ddp_params_and_buffers_to_ignore"):
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:604]: Curious why this change is needed?

File: torch/nn/parallel/distributed.py [hunk 6]
```diff
@@ -637,10 +640,10 @@ def __init__(
         parameters, expect_sparse_gradient = self._build_params_for_reducer()
         # Verify model equivalence.
         dist._verify_params_across_processes(self.process_group, parameters)
-        # Sync params and buffers. Ensures all DDP models start off at the same value.
-        self._sync_params_and_buffers(authoritative_rank=0)
         # In debug mode, build a mapping of parameter index -> parameter.
         param_to_name_mapping = self._build_debug_param_to_name_mapping(parameters)
+        # Sync params and buffers. Ensures all DDP models start off at the same value.
+        self._sync_params_and_buffers(authoritative_rank=0)
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:646]: Shall we only mark this after the sync completes successfully?

File: torch/nn/parallel/distributed.py [hunk 7]
```diff
@@ -665,6 +668,27 @@ def _sync_params_and_buffers(self, authoritative_rank=0):
                 module_states, self.broadcast_bucket_size, authoritative_rank
             )
 
+    def _mark_params_and_buffers_as_replicated(self, buffers_only=False):
+        from torch.distributed._shard.replicated_tensor import ReplicatedTensor, ReplicatedParameter
+        for module_name, module in self.module.named_modules():
+            if not buffers_only:
+                for param_name, param in module.named_parameters(recurse=False):
+                    if (param.requires_grad and f'{module_name}.{param_name}'
+                            not in self.parameters_to_ignore and not isinstance(param, ReplicatedParameter)):
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:677]: does the latter check mean that it is possible to call this function on already replicated parameters? Should we try to make it so that the application doesn't do this because it would be unnecessary?

File: torch/nn/parallel/distributed.py (hunk 2 above)
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:672]: Oops, I added `self.parameters_to_ignore` a couple of years ago and agree that a better name is `parameters_and_buffers_to_ignore`.

File: torch/nn/parallel/distributed.py [hunk 8]
```diff
@@ -665,6 +668,27 @@ def _sync_params_and_buffers(self, authoritative_rank=0):
...
+                        module.register_parameter(param_name, ReplicatedParameter(param))
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:678]: Do we not de-register the original, non-replicated parameters?

**Review by rohan-varma — COMMENTED** [review]
Are we considering enabling this change with a flag such as `DDP_USE_REPLICATED` to phase the rollout? I would advocate for this unless we're very confident about performance implications and reliability of `ReplicatedTensor`, as it would also have a large blast radius in OSS, and we may not be completely aware of it until the 1.12 release happens.

**Comment by pritamdamania87** [comment]
Closing in favor of: https://github.com/pytorch/pytorch/pull/75299

File: torch/distributed/_shard/replicated_tensor.py (hunk 1 above)
Comment (jbschlosser) [reply torch/distributed/_shard/replicated_tensor.py:144]: This is a bit delayed due to my medical leave, but wanted to leave an update here for posterity: #73459 is in now - please try it out if you'd like :)
